Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+   *
+   * Throws an error PLAYER_NOT_IN_GAME_ERROR if the current player is not a player in this game
+   */
+  get gamePiece(): 'X' | 'O'
+  /**
+   * Returns the status of the game.
+   */
+  get status(): GameStatus
+  /**
+   * Returns true if the game is in progress
+   */
+  public isActive(): boolean
+  /**
+   * Updates the internal state of this TicTacToeAreaController to match the new model.
+   *
+   * Calls super._updateFrom, which updates the occupants of this game area and
+   * other common properties (including this._model).
+   *
+   * If the board has changed, emits a 'boardChanged' event with the new board.
+   *
+   * If the turn has changed, emits a 'turnChanged' event with true if it is our turn, and false otherwise.
+   */
+  protected _updateFrom(newModel: GameArea<TicTacToeGameState>): void
+  /**
+   * Sends a request to the server to make a move in the game.
+   * Uses the this._townController.sendInteractableCommand method to send the request.
+   * The request should be of type 'GameMove',
+   * and send the gameID provided by `this._instanceID`.
+   *
+   * If the game is not in progress, throws an error NO_GAME_IN_PROGRESS_ERROR
+   *
+   * @param row Row of the move
+   * @param col Column of the move
+   */
+  public async makeMove(row: TicTacToeGridPosition, col: TicTacToeGridPosition) 
+{% endhighlight %}
+</details>
+
+*Testing*: Avery has provided you with tests for everything in `TicTacToeController` *except* for `makeMove` and `_updateFrom`. Please implement these additional tests in the file `frontend/src/classes/interactable/TicTacToeAreaController.test.ts`. You can run the tests by running the command `npx jest TicTacToeAreaController` in the `frontend` directory (for convenience, you may want to use `npx jest --watch` ...).
+
+The grading script will assign full marks for each implementation task if all of the tests for that task pass. The grading script will assign full marks for each testing task if all of the faults in that task are detected by your tests. There is no partial credit.
+
+Grading for implementation tasks:
+* All properties and methods besides `_updateFrom`: 10 points
+* `_updateFrom`: 10 points
+
+Grading for the testing tasks:
+* Tests for `_updateFrom`: 5 points
+* Tests for `makeMove`: 5 points
+
+### Task 2: Tic-Tac-Toe Area (50 points total)
+The next task is to implement the React component that will render the Tic-Tac-Toe game area. This component will show information about the game area, and the current state of the game. It displays the `TicTacToeBoard`` (which you'll implement in the next task), and also the `Leaderboard` component (which you'll implement in the task after that).
+
+This component is located in the file `frontend/src/components/Town/interactables/TicTacToe/TicTacToeArea.tsx` - you should implement component class in this file.
+
+<details><summary markdown="span">View the specification for this component</summary>
```

Comment:
With respect to this spec, my TA says:

> Could be helpful to mention that the elements should be from Chakra UI and not plain html. For example, the join game button (I probably would have started with plain html and then realized that it doesn’t have properties like `isLoading` or `disabled`)

> The specs don’t mention that the join game button should not appear if the status is
`WAITING_TO_START` and the player is already in the game. However, there is a test case for this
scenario so the specs can be determined from there.

> There was also some ambiguity with listing the players as ‘X: {username}’ and ‘O: {username}’ as
it did not specify needing the X: and O: in the specification, but again looking at the tests, it
should be more clear what is expected.

Reply:
Added small clarifications for all of these things, thanks!